File: rb/test/pull_entity_test.rb
# Pull entity test

require "minitest/autorun"
require "json"
require_relative "../GithubRest_sdk"
require_relative "runner"

class PullEntityTest < Minitest::Test
  def test_create_instance
    testsdk = GithubRestSDK.test(nil, nil)
    ent = testsdk.Pull(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = pull_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "pull." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set GITHUBREST_TEST_PULL_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    pull_ref01_ent = client.Pull(nil)
    pull_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.pull"), "pull_ref01"))
    pull_ref01_data["owner"] = setup[:idmap]["owner01"]
    pull_ref01_data["repo"] = setup[:idmap]["repo01"]

    pull_ref01_data_result, err = pull_ref01_ent.create(pull_ref01_data, nil)
    assert_nil err
    pull_ref01_data = Helpers.to_map(pull_ref01_data_result)
    assert !pull_ref01_data.nil?
    assert !pull_ref01_data["id"].nil?

    # LIST
    pull_ref01_match = {
      "owner" => setup[:idmap]["owner01"],
      "repo" => setup[:idmap]["repo01"],
    }

    pull_ref01_list_result, err = pull_ref01_ent.list(pull_ref01_match, nil)
    assert_nil err
    assert pull_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(pull_ref01_list_result),
      { "id" => pull_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # LOAD
    pull_ref01_match_dt0 = {
      "id" => pull_ref01_data["id"],
    }
    pull_ref01_data_dt0_loaded, err = pull_ref01_ent.load(pull_ref01_match_dt0, nil)
    assert_nil err
    pull_ref01_data_dt0_load_result = Helpers.to_map(pull_ref01_data_dt0_loaded)
    assert !pull_ref01_data_dt0_load_result.nil?
    assert_equal pull_ref01_data_dt0_load_result["id"], pull_ref01_data["id"]

  end
end

def pull_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "pull", "PullTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = GithubRestSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["pull01", "pull02", "pull03", "repo01", "repo02", "repo03", "owner01"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["GITHUBREST_TEST_PULL_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "GITHUBREST_TEST_PULL_ENTID" => idmap,
    "GITHUBREST_TEST_LIVE" => "FALSE",
    "GITHUBREST_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["GITHUBREST_TEST_PULL_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["GITHUBREST_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = GithubRestSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["GITHUBREST_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["GITHUBREST_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/pull_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-rest-sdk/go"
	"github.com/voxgig-sdk/github-rest-sdk/go/core"

	vs "github.com/voxgig-sdk/github-rest-sdk/go/utility/struct"
)

func TestPullEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Pull(nil)
		if ent == nil {
			t.Fatal("expected non-nil PullEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := pullBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "pull." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUBREST_TEST_PULL_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		pullRef01Ent := client.Pull(nil)
		pullRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "pull"}, setup.data), "pull_ref01"))
		pullRef01Data["owner"] = setup.idmap["owner01"]
		pullRef01Data["repo"] = setup.idmap["repo01"]

		pullRef01DataResult, err := pullRef01Ent.Create(pullRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		pullRef01Data = core.ToMapAny(pullRef01DataResult)
		if pullRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if pullRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		pullRef01Match := map[string]any{
			"owner": setup.idmap["owner01"],
			"repo": setup.idmap["repo01"],
		}

		pullRef01ListResult, err := pullRef01Ent.List(pullRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		pullRef01List, pullRef01ListOk := pullRef01ListResult.([]any)
		if !pullRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", pullRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(pullRef01List), map[string]any{"id": pullRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// LOAD
		pullRef01MatchDt0 := map[string]any{
			"id": pullRef01Data["id"],
		}
		pullRef01DataDt0Loaded, err := pullRef01Ent.Load(pullRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		pullRef01DataDt0LoadResult := core.ToMapAny(pullRef01DataDt0Loaded)
		if pullRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if pullRef01DataDt0LoadResult["id"] != pullRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func pullBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "pull", "PullTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read pull test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse pull test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"pull01", "pull02", "pull03", "repo01", "repo02", "repo03", "owner01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUBREST_TEST_PULL_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUBREST_TEST_PULL_ENTID": idmap,
		"GITHUBREST_TEST_LIVE":      "FALSE",
		"GITHUBREST_TEST_EXPLAIN":   "FALSE",
		"GITHUBREST_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["GITHUBREST_TEST_PULL_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITHUBREST_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["GITHUBREST_APIKEY"],
			},
			extra,
		})
		client = sdk.NewGithubRestSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUBREST_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUBREST_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
